CRM/Core/BAO/FinancialTrxn.php

```diff
@@ -696,4 +696,30 @@ public static function createDeferredTrxn($lineItems, $contributionDetails, $upd
     }
   }
 
+  /**
+   * Format Credit Card type.
+   *
+   * @param array $params
+   *
+   */
+  public static function formatCreditCardDetails(&$params) {
+    $creditCardType = CRM_Utils_Array::value('credit_card_type', $params);
+    if ($creditCardType) {
+      if (!is_numeric($creditCardType)) {
+        $creditCardTypes = CRM_Core_PseudoConstant::get('CRM_Financial_DAO_FinancialTrxn',
+          'card_type',
+          array('labelColumn' => 'name')
+        );
+        $params['card_type'] = array_search($creditCardType, $creditCardTypes);
+      }
+      else {
+        $params['card_type'] = $creditCardType;
+      }
+    }
+    $ccNumber = CRM_Utils_Array::value('credit_card_number', $params);
+    if ($ccNumber) {
+      $params['pan_truncation'] = substr($ccNumber, -4);
```

Comment:
I don't think CompleteTransaction, or any api that interacts with the db should ever send in a credit card number. 

(I would say any api but I do have an api that just does a payment in an extension, it doesn't write to the db, the controlling function does that. Actually it takes a token not a credit card number, but in theory arguably it could take a credit card number, although I'm not sure about that - would need to see an actual use case first)